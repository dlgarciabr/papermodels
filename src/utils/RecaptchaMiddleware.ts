import { MiddlewareNext, MiddlewareResponse, RequestMiddleware } from 'blitz';
import { IncomingMessage } from 'http';

interface IRequestMiddlewareRequest extends IncomingMessage {
  body: any;
  cookies: any;
  query: any;
}

interface IRequestMiddlewareResponse extends MiddlewareResponse {
  status: Function;
}

const protectedUrls = ['/api/rpc/getItemAnonymous', '/api/rpc/getItemsAnonymous'];

const validateCaptcha = async (gRecaptchaToken: string, res: IRequestMiddlewareResponse, next: MiddlewareNext) => {
  const gRecaptchaSecret = process.env.NEXT_RECAPTCHA_SECRET;
  if (!gRecaptchaToken || !gRecaptchaSecret) {
    res.status(500).json({
      status: 'failure',
      message: 'Error validating captcha'
    });
    return;
  }

  const reCaptchaResponse = await fetch('https://www.google.com/recaptcha/api/siteverify', {
    method: 'POST',
    headers: {
      'Content-Type': 'application/x-www-form-urlencoded'
    },
    body: `secret=${gRecaptchaSecret}&response=${gRecaptchaToken}`
  });

  const captchaResultJson = await reCaptchaResponse.json();
  if (captchaResultJson?.score <= 0.5) {
    res.status(200).json({
      status: 'failure',
      message: 'Google ReCaptcha Failure'
    });
  } else {
    await next();
  }
};

const RecaptchaMidleware: RequestMiddleware = async (
  request: IRequestMiddlewareRequest,
  response: IRequestMiddlewareResponse,
  next: MiddlewareNext
) => {
  if (protectedUrls.some((url) => url === request.url)) {
    await validateCaptcha(request.body.params.gRecaptchaToken, response, next);
  } else {
    await next();
  }
};

export default RecaptchaMidleware;
