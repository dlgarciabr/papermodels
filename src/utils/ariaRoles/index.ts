import { WIDGET } from './widget';
import { LANDMARK } from './landmark';
import { STRUCTURE } from './structure';

/**
 * Extracted from [a11y-types](https://github.com/natemoo-re/a11y-types),
 * all credits to [Nate Moore](https://github.com/natemoo-re)
 */
export const ARIA_ROLE = {
  /**
   * Widget roles act as standalone user interface widgets or as part of larger, composite widgets.
   */
  WIDGET,

  /**
   * Landmark Roles indicate regions of the page intended as navigational landmarks.
   */
  LANDMARK,
  /**
   * Structure roles describe structures that organize content in a page.
   *
   * Document structures are not usually interactive.
   */
  STRUCTURE
};
