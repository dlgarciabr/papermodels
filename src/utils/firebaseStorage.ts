import { FirebaseApp, FirebaseOptions, initializeApp, getApps } from 'firebase/app';
import { getStorage as getFirebaseStorage, listAll, ref, getDownloadURL } from 'firebase/storage';

const FILES_PATH = 'artifacts';

const getFirebaseApp = (): FirebaseApp => {
  const firebaseConfig: FirebaseOptions = {
    apiKey: 'AIzaSyDfH-Wfq2vzCT8ULhzWYbtv-6E3v3_Vw-s',
    authDomain: 'papermodels-5a9e1.firebaseapp.com',
    projectId: 'papermodels-5a9e1',
    storageBucket: 'papermodels-5a9e1.appspot.com',
    messagingSenderId: '427465803853',
    appId: '1:427465803853:web:fa782e8eb6f474f14df404',
    measurementId: 'G-B8W9NCLDY8'
  };
  if (!getApps().length) {
    // Initialize Firebase
    initializeApp(firebaseConfig);
  }
  return getApps()[0] as FirebaseApp;
};

export const getStorage = () => getFirebaseStorage(getFirebaseApp());

export const listAllFiles = async () => {
  const result = await listAll(ref(getStorage(), FILES_PATH));
  return result;
};

export const getFilePath = async (fileId) => {
  const url = await getDownloadURL(ref(getStorage(), `${FILES_PATH}/${fileId}`));
  return url;
};

export const saveImage = () => {
  throw 'Not implemented';
};

export const deleteImage = () => {
  throw 'Not implemented';
};
