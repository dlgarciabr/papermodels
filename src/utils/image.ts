/* istanbul ignore file -- @preserve */
// It's not possible to make a real test mocking document.createElement and Image
import { readAndCompressImage } from 'browser-image-resizer';

const generateThumbnailCanvas = (bytes: ArrayBuffer) =>
  new Promise<HTMLCanvasElement>((resolve) => {
    const downloadUrl = window.URL.createObjectURL(new Blob([bytes]));
    const img = document.createElement('img');
    img.onload = () => {
      const canvas = document.createElement('canvas');
      canvas.width = 100;
      canvas.height = 100;
      const ctx = canvas.getContext('2d');
      ctx?.drawImage(img, 0, 0, 100, 100);
      resolve(canvas);
    };
    img.src = downloadUrl;
  });

export const generateThumbnailBlobUrl = async (bytes: ArrayBuffer) => {
  const canvas = await generateThumbnailCanvas(bytes);
  return canvas.toDataURL();
};

export const generateThumbnailArrayBuffer = (bytes: ArrayBuffer) =>
  new Promise<ArrayBuffer>(async (resolve, reject) => {
    const canvas = await generateThumbnailCanvas(bytes);
    canvas.toBlob((blob) => {
      if (blob) {
        resolve(blob.arrayBuffer());
      } else {
        reject();
      }
    });
  });

const config = {
  quality: 0.9,
  width: 100,
  height: 100
};

export const compressImage = async (bytes: ArrayBuffer) => {
  const file = new File([bytes], 'temp');
  let resizedImage = await readAndCompressImage(file, config);
  return resizedImage;
};
