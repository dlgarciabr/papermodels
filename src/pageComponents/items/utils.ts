import { FileType } from 'db';
import { saveFile } from 'src/utils/fileStorage';
import { UploadItemFile } from './types';

export const uploadFiles = async (files: UploadItemFile[]) => {
  const promises: Promise<void>[] = [];
  files.forEach((file) => {
    file.artifactType = FileType.scheme; // TODO receive as parameter
    const index = ++file.item.files.length;
    const name = file.item.name.replaceAll(' ', '_').toLowerCase();
    const extension = file.name.split('.')[1];
    const storageName = `${file.item.id}/${name}_${file.artifactType}_${index}.${extension}`;
    file.storageName = storageName;
    promises.push(saveFile(file));
  });
  await Promise.all(promises);
};

export const saveItemFiles = async (files: UploadItemFile[], mutation: any) => {
  const promises: Promise<void>[] = [];
  files.forEach((file) => {
    promises.push(
      mutation({
        id: file.storageName,
        artifactType: file.artifactType,
        itemId: file.item.id
      })
    );
  });
  await Promise.all(promises);
};
