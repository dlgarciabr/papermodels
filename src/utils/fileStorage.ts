/* istanbul ignore file -- @preserve */
// TODO test coverage to be implemented (https://github.com/dlgarciabr/papermodels/issues/27)

import { UploadItemFile } from 'src/items/types';
import storageProvider from './storageProviders/cloudinary';

export const ARTIFACTS_PATH = process.env.NEXT_PUBLIC_STORAGE_ARTIFACTS_PATH || 'papermodel';

export const listAllFiles = async () => {
  throw new Error('not implemented');
};

// export const listAllFiles = async () => await storageProvider.listAllFiles();

/**
 * Retrieves a full downloadable artifact url from a pre-defined storage provider
 *
 * @param path initial stored artifact path
 * @returns full storage artifact url
 */
export const getFileUrl = (path: string) => storageProvider.getFileUrl(path);

export const getThumbnailUrl = (path: string) => storageProvider.getThumbnailUrl(path);

export const saveFile = (file: UploadItemFile) => storageProvider.saveFile(file, `${ARTIFACTS_PATH}/${file.item.id}`);

export const deleteFile = async (path: string) => storageProvider.deleteFile(path);
