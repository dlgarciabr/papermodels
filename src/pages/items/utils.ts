import { FileType, Item, ItemFile } from 'db';
import { deleteFile, getFilePath, saveFile } from 'src/utils/fileStorage';
import { compressImage, generateThumbnailArrayBuffer } from 'src/utils/image';
import { removeDiacritics } from 'src/utils/string';
import { UploadItemFile } from '../../items/types';

// TODO remover
export const testCallServerlessFunc = () => {
  console.log('##################testCallServerlessFunc############################');
};

export const processFiles = async (files: UploadItemFile[]) => {
  const processedFiles: UploadItemFile[] = [];
  for await (const file of files) {
    const index = ++file.item.files.length;
    const name = removeDiacritics(file.item.name).replaceAll(' ', '_').toLowerCase();
    const extension = file.name.split('.')[1];
    const storagePath = `${file.item.id}/${name}_${file.artifactType}_${index}`;
    const imageBytes = await file.arrayBuffer();

    /* istanbul ignore if @preserve */
    if (file.artifactType === FileType.preview) {
      const thumbFileName = `${storagePath}_thumb.${extension}`;
      const thumbnailBytes = await generateThumbnailArrayBuffer(imageBytes);
      const compressedThumbnail = await compressImage(thumbnailBytes);
      const thumbnailCompressedBytes = await compressedThumbnail.arrayBuffer();
      const thumbnailFile = new File([thumbnailCompressedBytes], thumbFileName) as UploadItemFile;
      thumbnailFile.artifactType = FileType.thumbnail;
      thumbnailFile.storagePath = thumbFileName;
      thumbnailFile.item = file.item;
      processedFiles.push(thumbnailFile);
    }

    const fileName = `${storagePath}.${extension}`;
    const processedFile = new File([imageBytes], fileName) as UploadItemFile;
    processedFile.artifactType = file.artifactType;
    processedFile.storagePath = fileName;
    processedFile.item = file.item;
    processedFiles.push(processedFile);
  }
  return processedFiles;
};

export const uploadFiles = (files: UploadItemFile[]) =>
  Promise.all(
    files.map(async (file) => {
      await saveFile(file);
    })
  );

export const saveItemFiles = async (files: UploadItemFile[], createFileMutation: any) => {
  for await (const file of files) {
    const index = ++file.item.files.length;
    await createFileMutation({
      storagePath: file.storagePath,
      artifactType: file.artifactType,
      itemId: file.item.id,
      index
    });
  }
};

export const sortFilesIndexes = async (
  item: Item,
  files: ItemFile[],
  updateItemFileMutation: any
): Promise<ItemFile[]> => {
  if (files.length === 0 || (files.length === 1 && files[0]?.index === 0)) {
    return files;
  }
  const lockFileStorageName = `${item.id}/.lock`;

  await saveFile(new File([], lockFileStorageName));

  const promises: Promise<void>[] = [];
  const updatedFiles: ItemFile[] = [];

  files.forEach((file, index) => {
    promises.push(
      new Promise(async (resolve) => {
        const backupFileName = `${file.storagePath}_bak`;
        const url = await getFilePath(file.storagePath);

        const response = await fetch(url, { method: 'GET' });
        const blob = await response.blob();

        await saveFile(new File([blob], backupFileName));
        await deleteFile(file.storagePath);

        const name = removeDiacritics(item.name).replaceAll(' ', '_').toLowerCase();
        const extension = file.storagePath.split('.')[1];
        const newStoragePath = `${item.id}/${name}_${file.artifactType}_${index + 1}.${extension}`;

        await saveFile(new File([blob], newStoragePath));
        await deleteFile(backupFileName);

        const updatedFile = await updateItemFileMutation({
          ...file,
          storagePath: newStoragePath,
          index
        });

        updatedFiles.push(updatedFile);
        resolve();
      })
    );
  });

  await Promise.all(promises);

  await deleteFile(lockFileStorageName);

  return updatedFiles;
};
