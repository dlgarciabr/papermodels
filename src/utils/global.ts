import { FileType, Item, ItemFile } from '.prisma/client';
import { getFilePath } from './fileStorage';

export const downloadFile = async (file: ItemFile & { url: string; item: Item }) => {
  const url = await getFilePath(file.storagePath);
  const response = await fetch(url, { method: 'GET' });
  const blob = await response.blob();

  // Create blob link to download
  // const downloadUrl = window.URL.createObjectURL(new Blob([blob]));
  const downloadUrl = window.URL.createObjectURL(blob);
  const link = document.createElement('a');
  link.href = downloadUrl;
  link.setAttribute('download', file.storagePath);

  // Append to html link element page
  document.body.appendChild(link);

  // Start download
  link.click();

  // Clean up and remove the link
  (link.parentNode as any).removeChild(link);
};

export const getSimpleRandomKey = () => Math.random().toString(36).substring(2, 15);

export const getFileTypeByText = (type: string): FileType => {
  switch (type) {
    case FileType.thumbnail:
      return FileType.thumbnail;
    case FileType.instruction:
      return FileType.instruction;
    case FileType.scheme:
      return FileType.scheme;
    case FileType.preview:
    default:
      return FileType.preview;
  }
};
