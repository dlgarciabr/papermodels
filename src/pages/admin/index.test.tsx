import { expect, vi } from 'vitest';

import { render } from 'test/utils';
import Home from './index.page';

test('Admin page is rendered', () => {
  // arrange
  vi.mock('src/users/hooks/useCurrentUser', () => ({
    useCurrentUser: () => ({
      id: 1,
      name: 'User',
      email: 'user@email.com',
      role: 'user'
    })
  }));

  // act
  const result = render(<Home />);

  // assert
  expect(result.baseElement).toMatchSnapshot();
});

test.todo('User click on logout buton');

test.todo('User click on login buton');
