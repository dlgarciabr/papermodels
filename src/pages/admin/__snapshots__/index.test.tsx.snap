// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Admin page is rendered 1`] = `
<body>
  <div>
    <div
      class="Toastify"
    />
    <div>
      <div
        class="container"
      >
        <main>
          <div
            class="logo"
          >
            <img
              alt="blitzjs"
              height="160px"
              layout="fixed"
              src="undefined"
              width="256px"
            />
          </div>
          <div
            class="buttons"
            style="margin-top: 1rem; margin-bottom: 1rem;"
          >
            <button
              class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium css-sghohy-MuiButtonBase-root-MuiButton-root"
              tabindex="0"
              type="button"
            >
              Logout
              <span
                class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
              />
            </button>
            <div>
              Hello 
              user@email.com
            </div>
          </div>
          <div>
            <p>
              <a
                href="/categories"
              >
                categories
              </a>
            </p>
            <p>
              <a
                href="/items"
              >
                items
              </a>
            </p>
            <p>
              <a
                href="/admin/integration"
              >
                integration
              </a>
            </p>
          </div>
        </main>
        <style>
          
          @import url('https://fonts.googleapis.com/css2?family=Libre+Franklin:wght@300;700&display=swap');

          html,
          body {
            padding: 0;
            margin: 0;
            font-family: 'Libre Franklin', -apple-system, BlinkMacSystemFont, Segoe UI, Roboto, Oxygen, Ubuntu,
              Cantarell, Fira Sans, Droid Sans, Helvetica Neue, sans-serif;
          }

          * {
            -webkit-font-smoothing: antialiased;
            -moz-osx-font-smoothing: grayscale;
            box-sizing: border-box;
          }
          .container {
            min-height: 100vh;
            display: flex;
            flex-direction: column;
            justify-content: center;
            align-items: center;
          }

          main {
            padding: 5rem 0;
            flex: 1;
            display: flex;
            flex-direction: column;
            justify-content: center;
            align-items: center;
          }

          main p {
            font-size: 1.2rem;
          }

          p {
            text-align: center;
          }

          footer {
            width: 100%;
            height: 60px;
            border-top: 1px solid #eaeaea;
            display: flex;
            justify-content: center;
            align-items: center;
            background-color: #45009d;
          }

          footer a {
            display: flex;
            justify-content: center;
            align-items: center;
          }

          footer a {
            color: #f4f4f4;
            text-decoration: none;
          }

          .logo {
            margin-bottom: 2rem;
          }

          .logo img {
            width: 300px;
          }

          .buttons {
            display: grid;
            grid-auto-flow: column;
            grid-gap: 0.5rem;
          }
          .button {
            font-size: 1rem;
            background-color: #6700eb;
            padding: 1rem 2rem;
            color: #f4f4f4;
            text-align: center;
          }

          .button.small {
            padding: 0.5rem 1rem;
          }

          .button:hover {
            background-color: #45009d;
          }

          .button-outline {
            border: 2px solid #6700eb;
            padding: 1rem 2rem;
            color: #6700eb;
            text-align: center;
          }

          .button-outline:hover {
            border-color: #45009d;
            color: #45009d;
          }

          pre {
            background: #fafafa;
            border-radius: 5px;
            padding: 0.75rem;
            text-align: center;
          }
          code {
            font-size: 0.9rem;
            font-family: Menlo, Monaco, Lucida Console, Liberation Mono, DejaVu Sans Mono, Bitstream Vera Sans Mono,
              Courier New, monospace;
          }

          .grid {
            display: flex;
            align-items: center;
            justify-content: center;
            flex-wrap: wrap;

            max-width: 800px;
            margin-top: 3rem;
          }

          @media (max-width: 600px) {
            .grid {
              width: 100%;
              flex-direction: column;
            }
          }
        
        </style>
      </div>
      <p>
        <a
          href="/items"
        >
          Back to List
        </a>
      </p>
    </div>
  </div>
</body>
`;
